make dev failed caused by sed

```
make dev
check-copyright
cd tools/check && GO111MODULE=on go build -mod=mod -o ../bin/mockgen github.com/golang/mock/mockgen
scripts/generate-mock.sh
cd tools/check && GO111MODULE=on go build -mod=mod -o ../bin/msgp github.com/tinylib/msgp
cd tools/check && GO111MODULE=on go build -mod=mod -o ../bin/stringer golang.org/x/tools/cmd/stringer
cd tools/check && GO111MODULE=on go build -mod=mod -o ../bin/mockery github.com/vektra/mockery/v2
go generate
cd tools/check && GO111MODULE=on go build -mod=mod -o ../bin/gofumports mvdan.cc/gofumpt
cd tools/check && GO111MODULE=on go build -mod=mod -o ../bin/shfmt mvdan.cc/sh/v3/cmd/shfmt
go: downloading mvdan.cc/sh/v3 v3.3.1
go: downloading github.com/google/renameio v1.0.1
go: downloading github.com/pkg/diff v0.0.0-20210226163009-20ebb0f2a09e
go: downloading mvdan.cc/editorconfig v0.2.0
cd tools/check && GO111MODULE=on go build -mod=mod -o ../bin/gci github.com/daixiang0/gci
go: downloading github.com/daixiang0/gci v0.10.1
run gci (format imports)
tools/bin/gci write $(find . -name '*.go' -type f | grep -vE 'vendor|_gen|proto|pb\.go|pb\.gw\.go|_mock.go') 2>&1 | awk '{ print } END { if (NR > 0) { exit 1  }  }'
run gofumports
tools/bin/gofumports -l -w $(find . -name '*.go' -type f | grep -vE 'vendor|_gen|proto|pb\.go|pb\.gw\.go|_mock.go') 2>&1 | awk '{ print } END { if (NR > 0) { exit 1  }  }'
run shfmt
tools/bin/shfmt -d -w .
check log style
scripts/check-log-style.sh
check-file-width
cd tools/check && GO111MODULE=on go build -mod=mod -o ../bin/golangci-lint github.com/golangci/golangci-lint/cmd/golangci-lint
go: downloading github.com/golangci/golangci-lint v1.52.2
go: downloading github.com/gofrs/flock v0.8.1
go: downloading github.com/hashicorp/go-version v1.6.0
go: downloading github.com/ldez/gomoddirectives v0.2.3
go: downloading github.com/go-xmlfmt/xmlfmt v1.1.2
go: downloading github.com/golangci/revgrep v0.0.0-20220804021717-745bb2f7c2e6
go: downloading 4d63.com/gocheckcompilerdirectives v1.2.1
go: downloading 4d63.com/gochecknoglobals v0.2.1
go: downloading github.com/Abirdcfly/dupword v0.0.11
go: downloading github.com/Antonboom/errname v0.1.9
go: downloading github.com/Antonboom/nilnil v0.1.3
go: downloading github.com/Djarvur/go-err113 v0.0.0-20210108212216-aea10b59be24
go: downloading github.com/GaijinEntertainment/go-exhaustruct/v2 v2.3.0
go: downloading github.com/OpenPeeDeeP/depguard v1.1.1
go: downloading github.com/alexkohler/prealloc v1.0.0
go: downloading github.com/alingse/asasalint v0.0.11
go: downloading github.com/ashanbrown/forbidigo v1.5.1
go: downloading github.com/bkielbasa/cyclop v1.2.0
go: downloading github.com/blizzy78/varnamelen v0.8.0
go: downloading github.com/bombsimon/wsl/v3 v3.4.0
go: downloading github.com/breml/bidichk v0.2.4
go: downloading github.com/breml/errchkjson v0.3.1
go: downloading github.com/butuzov/ireturn v0.1.1
go: downloading github.com/charithe/durationcheck v0.0.10
go: downloading github.com/curioswitch/go-reassign v0.2.0
go: downloading github.com/denis-tingaikin/go-header v0.4.3
go: downloading github.com/esimonov/ifshort v1.0.4
go: downloading github.com/firefart/nonamedreturns v1.0.4
go: downloading github.com/fzipp/gocyclo v0.6.0
go: downloading github.com/go-critic/go-critic v0.7.0
go: downloading github.com/golangci/check v0.0.0-20180506172741-cfe4005ccda2
go: downloading github.com/golangci/dupl v0.0.0-20180902072040-3e9179ac440a
go: downloading github.com/golangci/go-misc v0.0.0-20220329215616-d24fe342adfe
go: downloading github.com/golangci/gofmt v0.0.0-20220901101216-f2edd75033f2
go: downloading github.com/golangci/maligned v0.0.0-20180506175553-b1d89398deca
go: downloading github.com/golangci/misspell v0.4.0
go: downloading github.com/golangci/lint-1 v0.0.0-20191013205115-297bf364a8e0
go: downloading github.com/golangci/unconvert v0.0.0-20180507085042-28b1c447d1f4
go: downloading github.com/gordonklaus/ineffassign v0.0.0-20230107090616-13ace0543b28
go: downloading github.com/gostaticanalysis/nilerr v0.1.1
go: downloading github.com/jgautheron/goconst v1.5.1
go: downloading github.com/jirfag/go-printf-func-name v0.0.0-20200119135958-7558a9eaa5af
go: downloading github.com/julz/importas v0.1.0
go: downloading github.com/junk1tm/musttag v0.5.0
go: downloading github.com/kisielk/errcheck v1.6.3
go: downloading github.com/kkHAIKE/contextcheck v1.1.4
go: downloading github.com/kulti/thelper v0.6.3
go: downloading github.com/kunwardeep/paralleltest v1.0.6
go: downloading github.com/kyoh86/exportloopref v0.1.11
go: downloading github.com/ldez/tagliatelle v0.4.0
go: downloading github.com/leonklingele/grouper v1.1.1
go: downloading github.com/lufeee/execinquery v1.2.1
go: downloading github.com/maratori/testableexamples v1.0.0
go: downloading github.com/maratori/testpackage v1.1.1
go: downloading github.com/matoous/godox v0.0.0-20230222163458-006bad1f9d26
go: downloading github.com/mbilski/exhaustivestruct v1.2.0
go: downloading github.com/mgechev/revive v1.3.1
go: downloading github.com/moricho/tparallel v0.3.1
go: downloading github.com/nakabonne/nestif v0.3.1
go: downloading github.com/nishanths/exhaustive v0.9.5
go: downloading github.com/nunnatsa/ginkgolinter v0.9.0
go: downloading github.com/polyfloyd/go-errorlint v1.4.0
go: downloading github.com/ryancurrah/gomodguard v1.3.0
go: downloading github.com/ryanrolds/sqlclosecheck v0.4.0
go: downloading github.com/sanposhiho/wastedassign/v2 v2.0.7
go: downloading github.com/sashamelentyev/interfacebloat v1.1.0
go: downloading github.com/sashamelentyev/usestdlibvars v1.23.0
go: downloading github.com/securego/gosec/v2 v2.15.0
go: downloading github.com/shazow/go-diff v0.0.0-20160112020656-b6b7b6733b8c
go: downloading github.com/sivchari/containedctx v1.0.2
go: downloading github.com/sivchari/nosnakecase v1.7.0
go: downloading github.com/sivchari/tenv v1.7.1
go: downloading github.com/sonatard/noctx v0.0.2
go: downloading github.com/ssgreg/nlreturn/v2 v2.2.1
go: downloading github.com/stbenjam/no-sprintf-host-port v0.1.1
go: downloading github.com/sourcegraph/go-diff v0.7.0
go: downloading github.com/tdakkota/asciicheck v0.2.0
go: downloading github.com/tetafro/godot v1.4.11
go: downloading github.com/timakin/bodyclose v0.0.0-20221125081123-e39cf3fc478e
go: downloading github.com/timonwong/loggercheck v0.9.4
go: downloading github.com/tomarrell/wrapcheck/v2 v2.8.1
go: downloading github.com/tommy-muehle/go-mnd/v2 v2.5.1
go: downloading github.com/ultraware/funlen v0.0.3
go: downloading github.com/ultraware/whitespace v0.0.5
go: downloading github.com/uudashr/gocognit v1.0.6
go: downloading github.com/yagipy/maintidx v1.0.0
go: downloading github.com/yeya24/promlinter v0.2.0
go: downloading gitlab.com/bosi/decorder v0.2.3
go: downloading honnef.co/go/tools v0.4.3
go: downloading mvdan.cc/interfacer v0.0.0-20180901003855-c20040233aed
go: downloading mvdan.cc/unparam v0.0.0-20221223090309-7455f1af531d
go: downloading github.com/go-toolsmith/astfmt v1.1.0
go: downloading github.com/go-toolsmith/astcast v1.1.0
go: downloading github.com/go-toolsmith/astcopy v1.1.0
go: downloading github.com/go-toolsmith/astequal v1.1.0
go: downloading github.com/go-toolsmith/astp v1.1.0
go: downloading github.com/go-toolsmith/strparse v1.1.0
go: downloading github.com/go-toolsmith/typep v1.1.0
go: downloading github.com/quasilyte/go-ruleguard v0.3.19
go: downloading github.com/quasilyte/regex/syntax v0.0.0-20210819130434-b3f0c404a727
go: downloading github.com/ettle/strcase v0.1.1
go: downloading github.com/prometheus/client_golang v1.12.1
go: downloading golang.org/x/exp/typeparams v0.0.0-20230224173230-c95f2b4c22f2
go: downloading mvdan.cc/lint v0.0.0-20170908181259-adc824a0674b
go: downloading github.com/quasilyte/gogrep v0.5.0
go: downloading github.com/quasilyte/stdinfo v0.0.0-20220114132959-f7386bf02567
go: downloading github.com/chavacava/garif v0.0.0-20230227094218-b8c73b2037b8
go: downloading github.com/prometheus/common v0.32.1
go: downloading github.com/prometheus/procfs v0.7.3
tools/bin/golangci-lint run --timeout 10m0s --skip-dirs "^dm/","^tests/"
cd dm && ../tools/bin/golangci-lint run --timeout 10m0s
go mod tidy
./tools/check/check-tidy.sh
go: downloading github.com/creachadair/cityhash v0.1.1
go: downloading github.com/dgryski/go-spooky v0.0.0-20170606183049-ed3d087f40e2
go: downloading github.com/dgryski/go-stadtx v0.0.0-20181220175827-101ddca05f0b
go: downloading github.com/huichen/murmur v0.0.0-20130808212358-e0489551cf51
go: downloading google.golang.org/grpc/examples v0.0.0-20231221225426-4f03f3ff32c9
go: downloading github.com/leemcloughlin/gofarmhash v0.0.0-20160919192320-0a055c5b87a8
go: downloading github.com/pborman/uuid v1.2.1
go: downloading github.com/reusee/mmh3 v0.0.0-20140820141314-64b85163255b
go: downloading leb.io/hashland v0.1.5
go: downloading github.com/zhangxinngang/murmur v0.0.0-20140309145047-4e88ee1a5950
go: downloading golang.org/x/xerrors v0.0.0-20231012003039-104605ab7028
go: downloading github.com/containerd/cgroups v1.0.4
go: downloading github.com/lufia/plan9stats v0.0.0-20230326075908-cb1d2100619a
go: downloading github.com/power-devops/perfstat v0.0.0-20221212215047-62379fc7944b
go: downloading github.com/bytedance/sonic v1.9.1
go: downloading github.com/fortytw2/leaktest v1.3.0
go: downloading gopkg.in/check.v1 v1.0.0-20201130134442-10cb98267c6c
go: downloading github.com/cockroachdb/datadriven v1.0.2
go: downloading github.com/felixge/fgprof v0.9.3
go: downloading github.com/johannesboyne/gofakes3 v0.0.0-20230506070712-04da935ef877
go: downloading github.com/prometheus/prometheus v0.49.1
go: downloading github.com/sasha-s/go-deadlock v0.3.1
go: downloading github.com/otiai10/mint v1.3.1
go: downloading github.com/maxatome/go-testdeep v1.11.0
go: downloading google.golang.org/appengine v1.6.8
go: downloading github.com/godbus/dbus/v5 v5.0.4
go: downloading github.com/opencontainers/runtime-spec v1.0.2
go: downloading github.com/cilium/ebpf v0.4.0
go: downloading github.com/godbus/dbus v0.0.0-20190726142602-4481cbc300e2
go: downloading github.com/inconshreveable/mousetrap v1.1.0
go: downloading github.com/smartystreets/goconvey v1.6.4
go: downloading github.com/gavv/httpexpect v2.0.0+incompatible
go: downloading github.com/chzyer/test v1.0.0
go: downloading github.com/onsi/ginkgo v1.16.5
go: downloading github.com/onsi/gomega v1.20.1
go: downloading golang.org/x/lint v0.0.0-20201208152925-83fdc39ff7b5
go: downloading github.com/jcmturner/goidentity/v6 v6.0.1
go: downloading github.com/yusufpapurcu/wmi v1.2.3
go: downloading github.com/lib/pq v1.2.0
go: downloading github.com/tklauser/numcpus v0.6.1
go: downloading github.com/go-playground/assert/v2 v2.2.0
go: downloading github.com/cncf/xds/go v0.0.0-20231128003011-0fa0005c9caa
go: downloading github.com/jfcg/opt v0.3.1
go: downloading github.com/jfcg/rng v1.0.4
go: downloading github.com/petermattis/goid v0.0.0-20231207134359-e60b3f734c67
go: downloading github.com/ryszard/goskiplist v0.0.0-20150312221310-2dfbae5fcf46
go: downloading github.com/shabbyrobe/gocovmerge v0.0.0-20190829150210-3e036491d500
go: downloading github.com/decred/dcrd/dcrec/secp256k1/v4 v4.2.0
go: downloading github.com/HdrHistogram/hdrhistogram-go v1.1.2
go: downloading gonum.org/v1/gonum v0.8.2
go: downloading github.com/mwitkow/go-conntrack v0.0.0-20190716064945-2f068394615f
go: downloading github.com/go-errors/errors v1.0.1
go: downloading github.com/dnaeon/go-vcr v1.2.0
go: downloading github.com/dimfeld/httptreemux v5.0.1+incompatible
go: downloading github.com/frankban/quicktest v1.11.3
go: downloading cloud.google.com/go/pubsub v1.36.1
go: downloading github.com/google/martian/v3 v3.3.2
go: downloading github.com/jmespath/go-jmespath/internal/testify v1.5.1
go: downloading github.com/tidwall/assert v0.1.0
go: downloading github.com/tidwall/lotsa v1.0.2
go: downloading github.com/ajg/form v1.5.1
go: downloading github.com/google/go-querystring v1.0.0
go: downloading github.com/imkira/go-interpol v1.1.0
go: downloading github.com/moul/http2curl v1.0.0
go: downloading github.com/valyala/fasthttp v1.34.0
go: downloading github.com/yalp/jsonpath v0.0.0-20180802001716-5cc68e5049a0
go: downloading github.com/yudai/gojsondiff v1.0.0
go: downloading github.com/chzyer/logex v1.2.1
go: downloading github.com/jtolds/gls v4.20.0+incompatible
go: downloading github.com/smartystreets/assertions v0.0.0-20180927180507-b2de0cb4f26d
go: downloading github.com/go-ole/go-ole v1.3.0
go: downloading github.com/chenzhuoyu/base64x v0.0.0-20221115062448-fe3a3abad311
go: downloading golang.org/x/arch v0.3.0
go: downloading github.com/segmentio/asm v1.2.0
go: downloading github.com/envoyproxy/protoc-gen-validate v1.0.4
go: downloading github.com/shoenig/test v0.6.4
go: downloading github.com/alexbrainman/sspi v0.0.0-20210105120005-909beea2cc74
go: downloading github.com/jpillora/backoff v1.0.0
go: downloading github.com/danieljoos/wincred v1.1.2
go: downloading github.com/gsterjov/go-libsecret v0.0.0-20161001094733-a6f4afe4910c
go: downloading github.com/grafana/regexp v0.0.0-20221122212121-6b5c0a4cb7fd
go: downloading github.com/go-kit/log v0.2.1
go: downloading github.com/yudai/golcs v0.0.0-20170316035057-ecda9a501e82
go: downloading github.com/andybalholm/brotli v1.0.4
go: downloading github.com/gopherjs/gopherjs v0.0.0-20181017120253-0766667cb4d1
go: downloading github.com/nxadm/tail v1.4.8
go: downloading github.com/twitchyliquid64/golang-asm v0.15.1
go: downloading github.com/klauspost/cpuid/v2 v2.2.4
go: downloading github.com/dennwc/varint v1.0.0
go: downloading github.com/go-logfmt/logfmt v0.6.0
go: downloading github.com/fsnotify/fsnotify v1.7.0
check terror conflict
check pass
cd tools/check && GO111MODULE=on go build -mod=mod -o ../bin/errdoc-gen github.com/pingcap/errors/errdoc-gen
go: downloading github.com/pingcap/errors v0.11.5-0.20211009033009-93128226aaa3
generator errors.toml
# check-errdoc will skip DM directory.
./tools/check/check-errdoc.sh
Generate successfully, output:
Non-normalized error: proto: negative length found during unmarshaling
Non-normalized error: proto: integer overflow
Non-normalized error: proto: unexpected end of group
Non-normalized error: proto: negative length found during unmarshaling
Non-normalized error: proto: integer overflow
Non-normalized error: proto: unexpected end of group

check-merge-conflicts
check-ticdc-dashboard
check-file-width
/usr/bin/sed -i -e 's/^\(\t*\)  /\1\t/g' -e 's/^\(\t*\) /\1/' -- tools/Makefile Makefile engine/pkg/promutil/test/Makefile
sed: -e: No such file or directory
make: *** [format-makefiles] Error 1
```

--- a/.github/workflows/check_and_build.yaml
+++ b/.github/workflows/check_and_build.yaml
@@ -65,6 +65,9 @@ jobs:
       - name: Build
         run: make build
 
+      - name: Check
+        run: make check
+
   arm_build:
     runs-on: [ARM64]
     name: Arm Build

--- a/Makefile
+++ b/Makefile
@@ -51,11 +51,13 @@ endif
 # These logic is to check if the OS is Darwin, if so, add CGO_ENABLED=1.
 # ref: https://github.com/cloudfoundry/gosigar/issues/58#issuecomment-1150925711
 # ref: https://github.com/pingcap/tidb/pull/39526#issuecomment-1407952955
-OS    := "$(shell go env GOOS)"
+OS := "$(shell go env GOOS)"
 ifeq (${OS}, "linux")
 	CGO := 0
+	SED += -i
 else ifeq (${OS}, "darwin")
 	CGO := 1
+	SED += -i ''
 endif
 
 BUILD_FLAG =
@@ -150,7 +152,7 @@ check-makefiles: format-makefiles
 
 format-makefiles: ## Format all Makefiles.
 format-makefiles: $(MAKE_FILES)
-	$(SED) -i -e 's/^\(\t*\)  /\1\t/g' -e 's/^\(\t*\) /\1/' -- $?
+	$(SED) -e 's/^\(\t*\)  /\1\t/g' -e 's/^\(\t*\) /\1/' -- $?
 
 bank:
 	$(GOBUILD) -ldflags '$(LDFLAGS)' -o bin/bank ./tests/bank/bank.go ./tests/bank/case.go
@@ -527,7 +529,7 @@ dm_compatibility_test: check_third_party_binary_for_dm
 
 dm_coverage: tools/bin/gocovmerge tools/bin/goveralls
 	# unify cover mode in coverage files, more details refer to dm/tests/_utils/run_dm_ctl
-	find "$(DM_TEST_DIR)" -type f -name "cov.*.dmctl.*.out" -exec $(SED) -i "s/mode: count/mode: atomic/g" {} \;
+	find "$(DM_TEST_DIR)" -type f -name "cov.*.dmctl.*.out" -exec $(SED) "s/mode: count/mode: atomic/g" {} \;
 	tools/bin/gocovmerge "$(DM_TEST_DIR)"/cov.* | grep -vE ".*.pb.go|.*.pb.gw.go|.*.__failpoint_binding__.go|.*debug-tools.*|.*chaos.*" > "$(DM_TEST_DIR)/all_cov.out"
 	tools/bin/gocovmerge "$(DM_TEST_DIR)"/cov.unit_test*.out | grep -vE ".*.pb.go|.*.pb.gw.go|.*.__failpoint_binding__.go|.*debug-tools.*|.*chaos.*" > $(DM_TEST_DIR)/unit_test.out
 	go tool cover -html "$(DM_TEST_DIR)/all_cov.out" -o "$(DM_TEST_DIR)/all_cov.html"

--- a/cdc/scheduler/internal/v3/replication/replication_manager.go
+++ b/cdc/scheduler/internal/v3/replication/replication_manager.go
@@ -278,6 +278,7 @@ func (r *Manager) handleMessageHeartbeatResponse(
 			log.Info("schedulerv3: ignore table status no table found",
 				zap.String("namespace", r.changefeedID.Namespace),
 				zap.String("changefeed", r.changefeedID.ID),
+				zap.Any("from", from),
 				zap.Any("message", status))
 			continue
 		}
@@ -289,6 +290,7 @@ func (r *Manager) handleMessageHeartbeatResponse(
 			log.Info("schedulerv3: table has removed",
 				zap.String("namespace", r.changefeedID.Namespace),
 				zap.String("changefeed", r.changefeedID.ID),
+				zap.Any("from", from),
 				zap.Int64("tableID", status.Span.TableID))
 			r.spans.Delete(status.Span)
 		}

--- a/dm/tests/run_group.sh
+++ b/dm/tests/run_group.sh
@@ -5,28 +5,44 @@ set -eo pipefail
 CUR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 
 group=$1
+if [[ $group == "TLS_GROUP" ]]; then
+	group="G12"
+fi
+group_num=${group#G}
 
 # Define groups
 # Note: If new group is added, the group name must also be added to CI
 #  https://github.com/PingCAP-QE/ci/blob/main/pipelines/pingcap/tiflow/latest/pull_dm_integration_test.groovy
 # Each group of tests consumes as much time as possible, thus reducing CI waiting time.
 # Putting multiple light tests together and heavy tests in a separate group.
-declare -A groups
 groups=(
-	["G00"]="ha_cases_1 ha_cases_2 ha_cases2"
-	["G01"]="ha_cases3 ha_cases3_1 ha_master tracker_ignored_ddl validator_checkpoint incompatible_ddl_changes"
-	["G02"]="handle_error handle_error_2 handle_error_3 adjust_gtid async_checkpoint_flush binlog_parse case_sensitive checkpoint_transaction check_task"
-	["G03"]="dmctl_advance dmctl_basic dmctl_command downstream_diff_index downstream_more_column"
-	["G04"]="import_goroutine_leak incremental_mode initial_unit drop_column_with_index duplicate_event expression_filter extend_column fake_rotate_event"
-	["G05"]="load_interrupt many_tables online_ddl foreign_key full_mode gbk gtid ha_cases http_proxies"
-	["G06"]="relay_interrupt safe_mode sequence_safe_mode lightning_load_task lightning_mode metrics"
-	["G07"]="shardddl1 shardddl1_1 shardddl2 shardddl2_1"
-	["G08"]="shardddl3 shardddl3_1 shardddl4 shardddl4_1 sharding sequence_sharding sequence_sharding_removemeta"
-	["G09"]="import_v10x sharding2 ha new_collation_off only_dml openapi s3_dumpling_lightning sequence_sharding_optimistic"
-	["G10"]="start_task print_status http_apis new_relay all_mode"
+	# G00
+	"ha_cases_1 ha_cases_2 ha_cases2"
+	# G01
+	"ha_cases3 ha_cases3_1 ha_master tracker_ignored_ddl validator_checkpoint incompatible_ddl_changes"
+	# G02
+	"handle_error handle_error_2 handle_error_3 adjust_gtid async_checkpoint_flush binlog_parse case_sensitive checkpoint_transaction check_task"
+	# G03
+	"dmctl_advance dmctl_basic dmctl_command downstream_diff_index downstream_more_column"
+	# G04
+	"import_goroutine_leak incremental_mode initial_unit drop_column_with_index duplicate_event expression_filter extend_column fake_rotate_event"
+	# G05
+	"load_interrupt many_tables online_ddl foreign_key full_mode gbk gtid ha_cases http_proxies"
+	# G06
+	"relay_interrupt safe_mode sequence_safe_mode lightning_load_task lightning_mode metrics"
+	# G07
+	"shardddl1 shardddl1_1 shardddl2 shardddl2_1"
+	# G08
+	"shardddl3 shardddl3_1 shardddl4 shardddl4_1 sharding sequence_sharding sequence_sharding_removemeta"
+	# G09
+	"import_v10x sharding2 ha new_collation_off only_dml openapi s3_dumpling_lightning sequence_sharding_optimistic"
+	# G10
+	"start_task print_status http_apis new_relay all_mode"
 	# `others others_2 others_3` tests of old pipeline
-	["G11"]="validator_basic dm_syncer shardddl_optimistic slow_relay_writer sql_mode sync_collation"
-	["TLS_GROUP"]="tls"
+	# G11
+	"validator_basic dm_syncer shardddl_optimistic slow_relay_writer sql_mode sync_collation"
+	# G12 TLS_GROUP
+	"tls"
 )
 
 # Get other cases not in groups, to avoid missing any case
@@ -46,13 +62,12 @@ if [[ "$group" == "check others" ]]; then
 	fi
 	echo "Error: "$others" is not added to any group in dm/tests/run_group.sh"
 	exit 1
-elif [[ " ${!groups[*]} " =~ " ${group} " ]]; then
-	test_names="${groups[${group}]}"
+elif [[ $group_num =~ ^[0-9]+$ ]] && [[ -n ${groups[10#${group_num}]} ]]; then
+	# force use decimal index
+	test_names="${groups[10#${group_num}]}"
 	# Run test cases
-	if [[ -n $test_names ]]; then
-		echo "Run cases: ${test_names}"
-		"${CUR}"/run.sh "${test_names}"
-	fi
+	echo "Run cases: ${test_names}"
+	"${CUR}"/run.sh "${test_names}"
 else
 	echo "Error: invalid group name: ${group}"
 	exit 1

--- a/engine/test/integration_tests/run_group.sh
+++ b/engine/test/integration_tests/run_group.sh
@@ -5,19 +5,24 @@ set -eo pipefail
 CUR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 
 group=$1
+group_num=${group#G}
 
 # Define groups
 # Note: If new group is added, the group name must also be added to CI
 # https://github.com/PingCAP-QE/ci/blob/main/pipelines/pingcap/tiflow/latest/pull_engine_integration_test.groovy
 # Each group of tests consumes as much time as possible, thus reducing CI waiting time.
 # Putting multiple light tests together and heavy tests in a separate group.
-declare -A groups
 groups=(
-	["G00"]='dm_basic dm_case_sensitive dm_collation dm_dump_sync_mode'
-	["G01"]='dm_full_mode dm_lightning_checkpoint dm_many_tables'
-	["G02"]="dm_many_tables_local dm_new_collation_off dm_sql_mode"
-	["G03"]="dm_tls e2e_fast_finished e2e_node_failure e2e_with_selectors"
-	["G04"]="e2e_worker_error external_resource"
+	# G00
+	'dm_basic dm_case_sensitive dm_collation dm_dump_sync_mode'
+	# G01
+	'dm_full_mode dm_lightning_checkpoint dm_many_tables'
+	# G02
+	"dm_many_tables_local dm_new_collation_off dm_sql_mode"
+	# G03
+	"dm_tls e2e_fast_finished e2e_node_failure e2e_with_selectors"
+	# G04
+	"e2e_worker_error external_resource"
 )
 
 # Get other cases not in groups, to avoid missing any case
@@ -37,15 +42,14 @@ if [[ "$group" == "check others" ]]; then
 	fi
 	echo "Error: "$others" is not added to any group in engine/test/integration_tests/run_group.sh"
 	exit 1
-elif [[ " ${!groups[*]} " =~ " ${group} " ]]; then
-	test_names="${groups[${group}]}"
+elif [[ $group_num =~ ^[0-9]+$ ]] && [[ -n ${groups[10#${group_num}]} ]]; then
+	# force use decimal index
+	test_names="${groups[10#${group_num}]}"
 	# Run test cases
-	if [[ -n $test_names ]]; then
-		echo "Run cases: ${test_names}"
-		mkdir -p /tmp/tiflow_engine_test
-		"${CUR}"/run.sh "${test_names}" 2>&1 | tee /tmp/tiflow_engine_test/engine_it.log
-		./engine/test/utils/check_log.sh
-	fi
+	echo "Run cases: ${test_names}"
+	mkdir -p /tmp/tiflow_engine_test
+	"${CUR}"/run.sh "${test_names}" 2>&1 | tee /tmp/tiflow_engine_test/engine_it.log
+	./engine/test/utils/check_log.sh
 else
 	echo "Error: invalid group name: ${group}"
 	exit 1

--- a/tests/integration_tests/run_group.sh
+++ b/tests/integration_tests/run_group.sh
@@ -6,6 +6,7 @@ CUR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 
 sink_type=$1
 group=$2
+group_num=${group#G}
 
 # Other tests that only support mysql: batch_update_to_no_batch ddl_reentrant
 # changefeed_fast_fail changefeed_resume_with_checkpoint_ts sequence
@@ -28,35 +29,53 @@ storage_only_canal_json="canal_json_storage_basic canal_json_storage_partition_t
 # * https://github.com/PingCAP-QE/ci/blob/main/pipelines/pingcap/tiflow/latest/pull_cdc_integration_test.groovy
 # Each group of tests consumes as much time as possible, thus reducing CI waiting time.
 # Putting multiple light tests together and heavy tests in a separate group.
-declare -A groups
 groups=(
 	# Note: only the tests in the first three groups are running in storage sink pipeline.
-	["G00"]="$mysql_only $kafka_only $storage_only"
-	["G01"]="$mysql_only_http $kafka_only_protocol $storage_only_canal_json multi_tables_ddl"
-	["G02"]="$mysql_only_consistent_replicate $kafka_only_v2 $storage_only_csv"
-	["G03"]='row_format drop_many_tables processor_stop_delay partition_table'
-	["G04"]='foreign_key ddl_puller_lag ddl_only_block_related_table changefeed_auto_stop'
-	["G05"]='charset_gbk ddl_manager multi_source'
-	["G06"]='sink_retry changefeed_error ddl_sequence resourcecontrol'
-	# pulsar oauth2 authentication enabled
-	["G07"]='kv_client_stream_reconnect cdc split_region'
-	["G08"]='processor_err_chan changefeed_reconstruct multi_capture synced_status_with_redo'
-	["G09"]='gc_safepoint changefeed_pause_resume cli savepoint synced_status'
-	["G10"]='default_value simple cdc_server_tips event_filter sql_mode'
-	["G11"]='resolve_lock move_table autorandom generate_column'
-	["G12"]='many_pk_or_uk capture_session_done_during_task ddl_attributes'
-	# pulsar mtls authentication enabled
-	["G13"]='tiflash region_merge common_1'
-	["G14"]='big_txn changefeed_finish force_replicate_table'
-	["G15"]='new_ci_collation batch_add_table multi_rocks'
-	# currently G16 is not running in kafka pipeline
-	["G16"]='owner_resign processor_etcd_worker_delay sink_hang'
-	["G17"]='clustered_index processor_resolved_ts_fallback'
+	# G00
+	"$mysql_only $kafka_only $storage_only"
+	# G01
+	"$mysql_only_http $kafka_only_protocol $storage_only_canal_json multi_tables_ddl"
+	# G02
+	"$mysql_only_consistent_replicate $kafka_only_v2 $storage_only_csv"
+	# G03
+	'row_format drop_many_tables processor_stop_delay partition_table'
+	# G04
+	'foreign_key ddl_puller_lag ddl_only_block_related_table changefeed_auto_stop'
+	# G05
+	'charset_gbk ddl_manager multi_source'
+	# G06
+	'sink_retry changefeed_error ddl_sequence resourcecontrol'
+	# G07 pulsar oauth2 authentication enabled
+	'kv_client_stream_reconnect cdc split_region'
+	# G08
+	'processor_err_chan changefeed_reconstruct multi_capture synced_status_with_redo'
+	# G09
+	'gc_safepoint changefeed_pause_resume cli savepoint synced_status'
+	# G10
+	'default_value simple cdc_server_tips event_filter sql_mode'
+	# G11
+	'resolve_lock move_table autorandom generate_column'
+	# G12
+	'many_pk_or_uk capture_session_done_during_task ddl_attributes'
+	# G13 pulsar mtls authentication enabled
+	'tiflash region_merge common_1'
+	# G14
+	'big_txn changefeed_finish force_replicate_table'
+	# G15
+	'new_ci_collation batch_add_table multi_rocks'
+	# G16, currently G16 is not running in kafka pipeline
+	'owner_resign processor_etcd_worker_delay sink_hang'
+	# G17
+	'clustered_index processor_resolved_ts_fallback'
 	# only run the following tests in mysql pipeline
-	["G18"]='availability http_proxies sequence'
-	["G19"]='changefeed_fast_fail batch_update_to_no_batch changefeed_resume_with_checkpoint_ts'
-	["G20"]='tidb_mysql_test ddl_reentrant multi_cdc_cluster'
-	["G21"]='bank kill_owner_with_ddl owner_remove_table_error'
+	# G18
+	'availability http_proxies sequence'
+	# G19
+	'changefeed_fast_fail batch_update_to_no_batch changefeed_resume_with_checkpoint_ts'
+	# G20
+	'tidb_mysql_test ddl_reentrant multi_cdc_cluster'
+	# G21
+	'bank kill_owner_with_ddl owner_remove_table_error'
 )
 
 # Get other cases not in groups, to avoid missing any case
@@ -76,13 +95,12 @@ if [[ "$group" == "others" ]]; then
 	fi
 	echo "Error: "$others" is not added to any group in tests/integration_tests/run_group.sh"
 	exit 1
-elif [[ " ${!groups[*]} " =~ " ${group} " ]]; then
-	test_names="${groups[${group}]}"
+elif [[ $group_num =~ ^[0-9]+$ ]] && [[ -n ${groups[10#${group_num}]} ]]; then
+	# force use decimal index
+	test_names="${groups[10#${group_num}]}"
 	# Run test cases
-	if [[ -n $test_names ]]; then
-		echo "Run cases: ${test_names}"
-		"${CUR}"/run.sh "${sink_type}" "${test_names}"
-	fi
+	echo "Run cases: ${test_names}"
+	"${CUR}"/run.sh "${sink_type}" "${test_names}"
 else
 	echo "Error: invalid group name: ${group}"
 	exit 1

